system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the JS file backend and JS file of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "wasmfs.h"
+
+using js_index_t = uint32_t;
+
+extern "C" {
+void _emscripten_write_js_file(js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset);
+void _emscripten_read_js_file(js_index_t index,
+                              const uint8_t* buffer,
+                              size_t length,
+                              off_t offset);
+int _emscripten_get_js_file_size(js_index_t index);
+int _emscripten_create_js_file();
+void _emscripten_remove_js_file(js_index_t index);
+}
+
+namespace wasmfs {
+
+// This class describes a file that lives in JS Memory
+class JSFile : public DataFile {
+  // This index indicates the location of the JS File in the backing JS array.
+  js_index_t index;
+
+  // JSFiles will write from a Wasm Memory buffer into the backing JS array.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+
+    _emscripten_write_js_file(index, buf, len, offset);
+
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  // JSFiles will read from the backing JS array into a Wasm Memory buffer.
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    assert(offset + len - 1 < getSize());
```

Comment:
Assertions make sense for internal logic during testing. In production one would ship with assertions disabled and things should still be fine. But this is a check that depends on user input and the user may provide invalid input, which we should error on even in a release build.

Reply:
This assertion is just for internal logic. I check that we have not exceeded the file size in the caller. This is just an additional assertion during testing.

```
ssize_t dataLeft = size - currOffset;
    if (dataLeft <= 0) {
      break;
    }
```

Reply:
Ah, I see the comment now, sounds good.

Reply:
```suggestion
    assert(offset + len <= getSize());
```